### PR TITLE
Dev/robin/8241 accumulator prove and verify

### PR DESCRIPTION
**Pull request by robinbryce**
MMR draft alignment changes

The primary change here is to make our consistency proofs and inclusion proofs work in terms of the accumulator defined in the paper referenced by the COSE Receipts MMRIVER [draft](https://www.ietf.org/archive/id/draft-bryce-cose-merkle-mountain-range-proofs-00.html). And generaly include full support for [COSE Recieipts](https://cose-wg.github.io/draft-ietf-cose-merkle-tree-proofs/draft-ietf-cose-merkle-tree-proofs.html)

It also cleans up various algorithm implementations and checks that the draft algorithm implmentations and our go lang immplementation both meet the test vectors published by the draft.

Because it remains useful to be able to do so, I have retained the original variants of algorithms that worked with mmr "bagged" peaks. Some algorithm implementations I think we can remove altogether, but in this pr have not been deleted. Instead they are used in tests which show the "old" versions match the "new" versions

So in summary

* <file>bagged.go means a file that has things only relevant for MMR bagged peaks
* <file>old.go means a file that has things that are retained for compatibility testing but which are now "depricated"
* <func>Bagged() means the function is only relevant for working with MMR bagged peaks
* <func>Old() means the function is depricated, retained for compatibility testing

I've run all the local unit and integration tests. I've also run the integration tests in forestrie and checked that the arrangements for dealing with the change in consistency proofs work at that level. More testing for this on the forestrie side will be required before taking a release of go-datatrails-merklelog/mmr which has these changes.

### CODE REVIEW

File: massifs/go.mod [hunk 1]
```diff
@@ -2,6 +2,7 @@ module github.com/datatrails/go-datatrails-merklelog/massifs
 
 go 1.22
 
+replace github.com/datatrails/go-datatrails-merklelog/mmr => ../mmr
```
Comment (honourfish) [inline massifs/go.mod:5]: minor: we should remove this replace

I did this accidentally before, the solution is once this PR is merged to release mmr and not massifs, then change massifs to depend on the new mmr version in a 1 line PR, once that is merged we can release massifs

File: mmr/peaksold.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+//	3            15
+//	           /    \
+//	          /      \
+//	         /        \
+//	2       7          14
+//	      /   \       /   \
+//	1    3     6    10     13      18
+//	    / \  /  \   / \   /  \    /  \
+//	0  1   2 4   5 8   9 11   12 16   17
+func PeaksOld(mmrSize uint64) []uint64 {
```
Comment (honourfish) [inline mmr/peaksold.go:32]: I think we can make this more managable, either by completely removing old code (we have it released and versioned) or moving it to its own module that is better named than old?

Comment (robinbryce) [reply mmr/peaksold.go:32]: The plan is to delete everything called 'old' as soon as the updates are stable in prod.

For the "Bagged" support, it is possible we can get rid of that too but it would require a migration of existing seals. This is less "hard" than it sounds but its still something that could easily make replicated logs look "broken" even though the are provably good.

File: massifs/go.mod (hunk 1 above)
Comment (robinbryce) [reply massifs/go.mod]: ah got it. so we merge. then we do tiny follow up release + re-merge. thats fine I can do that

File: massifs/heightindex.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	return (1 << height) - uint64(height) - 1
+}
+
+// TreeSize returns the maximum byte size of the tree based on the defined log
+// entry size
+func TreeSize(height uint8) uint64 {
+	return TreeCount(height) * LogEntryBytes
+}
+
+// MaxCount returns the node count
```
Comment (honourfish) [inline massifs/heightindex.go:33]:
```suggestion
//  TreeCount returns the node count
```

File: massifs/heightindex.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package massifs
+
+// TreeRootIndex returns the root index for the tree with height
+func TreeRootIndex(height uint8) uint64 {
+	return (1 << height) - 2
+}
+
+// RangeRootIndex return the Massif root node's index in the overall MMR  given
```
Comment (honourfish) [inline massifs/heightindex.go:8]:
```suggestion
// RangeRootIndex return the Massif root node's mmr index in the overall MMR  given
```

File: massifs/massifcontextverified.go [hunk 5]
```diff
@@ -213,24 +229,26 @@ func (mc *MassifContext) verifyContext(
...
+				mmr.ErrConsistencyCheck)
 		}
 	}
 
 	return &VerifiedContext{
 		MassifContext:   *mc,
 		Sign1Message:    *msg,
 		MMRState:        state,
-		ConsistentRoots: rootB,
+		ConsistentRoots: peaksB,
```
Comment (honourfish) [inline massifs/massifcontextverified.go:252]: I think there are alot of terms floating around for what is basically the accumulator? Would it be worth adding an entry to the cheatsheet and in a separate PR making the names consistent.

Things like peakBagRHS, ConsistentRoots, peaksA, peaksB...etc (i know consistent roots are not always accumulators by their definition)

File: massifs/mmriver.go [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+
+	commoncose "github.com/datatrails/go-datatrails-common/cose"
+	"github.com/datatrails/go-datatrails-common/logger"
+	"github.com/datatrails/go-datatrails-merklelog/mmr"
+)
+
+// MMRIVER COSE Receipts to accompany our COSE MMRIVER seals
+
+type MMRiverInclusionProof struct {
+	Index         uint64   `cbor:"1,keyasint"`
```
Comment (honourfish) [inline massifs/mmriver.go:21]: love the use of `keyasint` much clearer as we can call the struct field a semantic name instead of a number

File: massifs/mmriver.go [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+	InclusionPath [][]byte `cbor:"2,keyasint"`
+}
+
+type MMRiverConsistencyProof struct {
+	TreeSize1        uint64   `cbor:"1,keyasint"`
+	TreeSize2        uint64   `cbor:"2,keyasint"`
+	ConsistencyPaths [][]byte `cbor:"3,keyasint"`
+	RightPeaks       [][]byte `cbor:"4,keyasint"`
```
Comment (honourfish) [inline massifs/mmriver.go:29]: minor: again in terms of accumulator term.

Personally I really like the term accumulator. 

the fact that it is the list of the peaks from the rhs is a nice property of mmrs?

File: massifs/mmriver.go [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+	ConsistencyProofs []MMRiverConsistencyProof `cbor:"-2,keyasint,omitempty"`
+}
+
+// MMRiverInclusionProofHeader provides for encoding, and defered decoding, of
+// COSE_Sign1 message headers for MMRIVER receipts
+type MMRiverVerifiableProofsHeader struct {
+	VerifiableProofs MMRiverVerifiableProofs `cbor:"396,keyasint"`
+}
+
+/*
```
Comment (honourfish) [inline massifs/mmriver.go:43]: minor: commented out code needs context or deletion?

File: massifs/mmriver.go [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+// VerifySignedInclusionReceipts verifies a signed COSE receipt encoded according to the MMRIVER VDS
+// on success the produced root is returned.
+// Signature verification failure is not an error, but the returned root will be nil and the result will be false.
+// All other unexpected issues are returned as errors, with a false result and nil root.
+// Note that MMRIVER receipts allow for multiple inclusion proofs to be attached to the receipt.
+// This function returns true only if ALL receipts verify
+func VerifySignedInclusionReceipts(
+	ctx context.Context,
+	receipt *commoncose.CoseSign1Message,
+	candidates [][]byte,
```
Comment (honourfish) [inline massifs/mmriver.go:58]: minor: can we add more context to what a `candidate` is?

File: massifs/mmriver.go [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+
+	var proof MMRiverInclusionProof
+
+	proof = verifiableProofs.InclusionProofs[0]
+	receipt.Payload = mmr.IncludedRoot(
+		sha256.New(),
+		proof.Index, candidates[0],
+		proof.InclusionPath)
+
+	err = receipt.VerifyWithCWTPublicKey(nil)
```
Comment (honourfish) [inline massifs/mmriver.go:94]: minor: why not verify the signature of the receipt before checking the inclusion root of the first candidate?

Comment (honourfish) [reply massifs/mmriver.go:94]: ahh because we need to set the payload in order to get the correct pre-image for the signature to verify. 

I get it!

File: massifs/mmriver.go [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+			return false, nil, fmt.Errorf(
+				"MMRIVER receipt VERIFY FAILED for: mmrIndex %d, candidate %d, err %v", proof.Index, i, err)
+		}
+	}
+	return true, receipt.Payload, nil
+}
+
+// VerifySignedInclusionReceipt verifies a reciept comprised of a single inclusion proof
+// If there are 0 or more than 1 candidates, the result will be false and an error will be returned
+func VerifySignedInclusionReceipt(
```
Comment (honourfish) [inline massifs/mmriver.go:115]: Fantastic!

File: massifs/mmriver.go [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+}
+
+type verifiedContextGetter interface {
+	GetVerifiedContext(
+		ctx context.Context, tenantIdentity string, massifIndex uint64,
+		opts ...ReaderOption,
+	) (*VerifiedContext, error)
+}
+
+// GetReceipt returns a COSE receipt for the given tenantIdentity and mmrIndex
```
Comment (honourfish) [inline massifs/mmriver.go:138]:
```suggestion
// NewReceipt returns a COSE receipt for the given tenantIdentity and mmrIndex
```

File: massifs/mmriver.go [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+func NewReceipt(
+	ctx context.Context,
+	massifHeight uint8,
+	tenantIdentity string, mmrIndex uint64,
```
Comment (honourfish) [inline massifs/mmriver.go:142]: minor: should we start using the term logTenant?

File: massifs/mmriver.go [hunk 14]
```diff
@@ -0,0 +1,246 @@
...
+		return false, nil, fmt.Errorf(
+			"MMRIVER receipt VERIFY FAILED for: mmrIndex %d, candidate %d, err %v", proof.Index, 0, err)
+	}
+	// verify the first proof then just compare the produced roots
+
+	for i := 1; i < len(verifiableProofs.InclusionProofs); i++ {
+
+		proof = verifiableProofs.InclusionProofs[i]
+		proven := mmr.IncludedRoot(sha256.New(), proof.Index, candidates[i], proof.InclusionPath)
+		if bytes.Compare(receipt.Payload, proven) != 0 {
```
Comment (honourfish) [inline massifs/mmriver.go:105]: This means that there is a constraint on all the inclusion paths to always lead to the same accumulator/root?

Which shouldn't be an issue, but i see further down we currently only create a receipt with 1 candidate.

Is it something we wish to do in the future to create receipts with multiple candidates?

File: massifs/rootsigner.go [hunk 15]
```diff
@@ -85,29 +158,159 @@ func (rs RootSigner) Sign1(coseSigner cose.Signer, keyIdentifier string, publicK
...
+// Note: regarding why and how we can pre-sign receipts:
+//
+// A specific advantage of MMR's is that we can pre-sign the protected headers
+// for all receipts we will ever be asked for. The scitt endpoint then only has
+// to copy the pre-signed receipt and *add* the inclusion path it is asked for.
+//
+// Importantly, this allows for self service *privacy preserving*, scitt
+// compatible, receipts based on replicated copies of the log.
+//
+// The most natural place to produce the pre-signed receipts is in the the log
```
Comment (honourfish) [inline massifs/rootsigner.go:189]: Is it worth mentioning that the payload of the cose sign 1 message receipt is the accumulator peak for any particular mmr state?

I found that is pretty cool

Comment (honourfish) [reply massifs/rootsigner.go:189]: maybe a copy pasta from the comment below?

File: mmr/consistentroots.go [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+func ConsistentRoots(hasher hash.Hash, ifrom uint64, accumulatorfrom [][]byte, proofs [][][]byte) ([][]byte, error) {
+	frompeaks := Peaks(ifrom)
+
+	if len(frompeaks) != len(proofs) {
+		return nil, ErrAccumulatorProofLen
+	}
+
+	roots := [][]byte{}
+
+	for iacc := 0; iacc < len(accumulatorfrom); iacc++ {
```
Comment (honourfish) [inline mmr/consistentroots.go:39]: minor: not a big fan of `iacc` which i think means `index accumulator`? but its very minor.

personal preference on i or accumulatorIndex

**Review by honourfish — DISMISSED** [review]

File: massifs/massifcontextverified.go (hunk 5 above)
Comment (robinbryce) [reply massifs/massifcontextverified.go:252]: Yeah, I think we do need to do that pass. I think a mmriver redux pr is in order to follow up.

File: massifs/mmriver.go (hunk 9 above)
Comment (robinbryce) [reply massifs/mmriver.go:58]: It is the value we believe is a "leaf" but untile we prove that it is a candidateLeaf

File: massifs/mmriver.go (hunk 13 above)
Comment (robinbryce) [reply massifs/mmriver.go:142]: yes. definitely. I might do that as a follow up in the 'redux' pr as there will be some references to be careful of

File: massifs/mmriver.go (hunk 14 above)
Comment (robinbryce) [reply massifs/mmriver.go:105]: Yes, a single receipt can contain proofs for many leaves. But only if they share the same accumulator peak. The longer the leaf is in the log, the more siblings it will get. That came for free and has little divergence with more traditional CT style receipts.

We _could_ do a full multi receipt where the payload is a catenation of leaves, but I think that we should wait for a bit before considering that. If I put that in the MMRIVER draft it will almost certainly have some negative feedback.

File: massifs/rootsigner.go (hunk 15 above)
Comment (robinbryce) [reply massifs/rootsigner.go:183]: I've updated the comment in signEmptyPeakReceipts to be more explanitory, including highlighting that very important "low update frequency" property for the peaks & proofs

File: massifs/mmriver.go (hunk 11 above)
Comment (robinbryce) [reply massifs/mmriver.go:112]: ta :-)

File: massifs/mmriver.go (hunk 9 above)
Comment (robinbryce) [reply massifs/mmriver.go:55]: added some commentary

File: massifs/mmriver.go (hunk 8 above)
Comment (robinbryce) [reply massifs/mmriver.go]: done

**Review by honourfish — APPROVED** [review]
